src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -109,6 +132,56 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
 }
 
 
+/* TODO: write comment */
+static void
+InitTransactionStateForTask(Task *task)
+{
+	MemoryContext oldContext = NULL;
+	ListCell *placementCell = NULL;
+
+	oldContext = MemoryContextSwitchTo(TopTransactionContext);
+
+	foreach(placementCell, task->taskPlacementList)
+	{
+		ShardPlacement *placement = (ShardPlacement *) lfirst(placementCell);
+		PGconn *connection = GetOrEstablishConnection(placement->nodeName,
+													  placement->nodePort);
+		TxnParticipant *txnParticipant = NULL;
+		bool participantHealthy = true;
+		txnParticipant = (TxnParticipant *) palloc(sizeof(TxnParticipant));
+
+		CopyShardPlacement(placement, (ShardPlacement *) txnParticipant);
+		txnParticipant->connection = connection;
+
+		if (txnParticipant->connection != NULL)
+		{
+			PGresult *result = PQexec(connection, "BEGIN");
```

Comment:
Hm. If there's a number of placements, the cost of the roundtrip could be mitigated by sending the BEGIN asynchronously to each, and only waiting at the end. The current way scales linearly with the number of placements, the other near constant (if looking at RTT).  Not sure if it has to be done now, but it seems worth mentioning.